crates/build/re_types_builder/src/docs.rs

```diff
@@ -15,103 +17,430 @@ impl Docs {
+                    continue;
                 }
-            })
-            .collect();
 
-        // NOTE: remove duplicated blank lines.
-        lines.dedup();
+                out_tokens.push(translate_doclink(&doclink_tokens, target));
+                continue;
+            }
+
+            // Normal boring token
+            out_tokens.push(token.to_owned());
+        }
+
+        out_tokens.into_iter().collect()
+    }
+
+    fn translate_doclink(doclink_tokens: &[&str], target: Target) -> String {
+        try_translate_doclink(doclink_tokens, target).unwrap_or_else(|err| {
+            let original_doclink: String = doclink_tokens.join("");
+
+            // The worlds simplest heuristic, but at least it doesn't warn about things like [x, y, z, w].
+            let looks_like_rerun_doclink =
+                !original_doclink.contains(' ') && original_doclink.len() > 6;
+
+            if looks_like_rerun_doclink {
+                re_log::warn_once!(
+                    "Looks like a Rerun doclink, but fails to parse: {original_doclink} - {err}"
+                );
+            }
+
+            original_doclink
+        })
+    }
+
+    fn try_translate_doclink(
+        doclink_tokens: &[&str],
+        target: Target,
+    ) -> Result<String, &'static str> {
+        let mut tokens = doclink_tokens.iter();
+        if tokens.next() != Some(&"[") {
+            return Err("Missing opening bracket");
+        }
+        let kind = *tokens.next().ok_or("Missing kind")?;
+        if kind == "`" {
+            return Err("Do not use backticks inside doclinks");
+        }
+        if tokens.next() != Some(&".") {
+            return Err("Missing dot");
+        }
+        let type_name = tokens.next().ok_or("Missing type name")?;
+        if tokens.next() != Some(&"]") {
+            return Err("Missing closing bracket");
+        }
+        if tokens.next().is_some() {
+            return Err("Trailing tokens");
+        }
+
+        // NOTE: we don't do any validation that the target exists.
+        // Instead we rely on the documentation tools for the different targets,
+        // e.g. `cargo doc` and our url link checker.
+
+        Ok(match target {
+            Target::Cpp => format!("`{kind}::{type_name}`"),
+            Target::Rust => {
+                // https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html
+                format!("[`{kind}::{type_name}`][crate::{kind}::{type_name}]")
+            }
+            Target::Python => format!("[`{kind}.{type_name}`][rerun.{kind}.{type_name}]"),
+            Target::WebDocsMarkdown => {
+                // For instance, https://rerun.io/docs/reference/types/views/spatial2d_view
+                let type_name_snake_case = re_case::to_snake_case(type_name);
+                format!("[`{kind}.{type_name}`](https://rerun.io/docs/reference/types/{kind}/{type_name_snake_case})")
```

Comment:
Can we make these links not point to `rerun.io`, but instead point to the correct file in under `/docs` as a relative path? As it is, you can't use these links to navigate between files. `rerun.io` will do the right thing if these are turned into relative file paths.

Reply:
> As it is, you can't use these links to navigate between files. 

What do you mean by this? Surely clicking the links will work, no? Will it reset some state that we want kept?

---

Relative path (`../archetypes/image`) has the drawback that the link we generate depends on where we are, which is super annoying. Would domain-relative paths (`/docs/reference/types/archetypes/image`) work instead?

---

Having these as absolute (`rerun.io/…`) paths means we can use them in the markdown that we embed in the viewer, so if we change that we'll have to add two different `Target`s (e.g. `RerunIoMarkdownWithRelativeLinks` and `MarkdownDocs`)

Reply:
I mean if you open the repository in vscode, navigate to docs/content/some_file.md, and click a link, it will take you to rerun.io instead of opening a different file in the editor. It seems like vscode treats those "domain-relative" paths as relative to the workspace root, so they should work.

The markdown not being embeddable in the viewer is a pre-existing problem because we have tons of relative links in there already. For example, look at https://github.com/rerun-io/rerun/blob/main/docs/content/reference/types/archetypes/annotation_context.md, and click the link to the `AnnotationContext` component. How would that link work in the viewer? That issue needs to be solved but it can be done separately

Reply:
> The markdown not being embeddable in the viewer is a pre-existing problem

It's pre-existing to embed the docs generated by `codegen/docs`, but we already show the docstring of components when you hover them in the viewer, and that's handled by `crates/build/re_types_builder/src/codegen/rust/reflection.rs`, so it is two separate (but related) things.

Reply:
Sure, my concern here is not blocking, just wanted to see if you think it was possible or not. If it's too annoying then we can just use `rerun.io` links.